### PR TITLE
fix: allow -Wpedantic in C++20 mode

### LINKED ISSUE #5310: [BUG]:  token pasting of ',' and __VA_ARGS__ is a GNU extension / pedantic warnings
### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [ ] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### What version (or hash if on master) of pybind11 are you using?

2.13.3

### Problem description

The most recent release uses `, ##__VA_ARGS__` in `common.h`. This is not technically in the standard (see e.g. [here](https://en.cppreference.com/w/cpp/preprocessor/replace#.23_and_.23.23_operators)), but most compilers support it. Clang, however, complains about its use when setting `-Wpedantic`. I can work around this by either not passing `-Wpedantic`, or passing `-Wno-gnu-zero-variadic-macro-arguments`, but I'd rather not do that if it can at all be avoided.

Consider the attached example, `test.cpp`. Compiling it with 
```shell
clang -Wall -Wextra -Wpedantic -Werror -shared -std=c++20 -fPIC $(python -m pybind11 --includes) test.cpp -o test.so
```
produces the following:
```shell
In file included from test.cpp:1:
In file included from .venv/lib/python3.11/site-packages/pybind11/include/pybind11/pybind11.h:13:
In file included from .venv/lib/python3.11/site-packages/pybind11/include/pybind11/detail/class.h:12:
In file included from .venv/lib/python3.11/site-packages/pybind11/include/pybind11/attr.h:13:
.venv/lib/python3.11/site-packages/pybind11/include/pybind11/detail/common.h:493:57: error: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Werror,-Wgnu-zero-variadic-macro-arguments]
            &PYBIND11_CONCAT(pybind11_module_def_, name),
  \
                                                        ^
1 error generated.
```
It concerns the following code snippet:
https://github.com/pybind/pybind11/blob/fc97cc41d575d44fa425a0e56765380e15474399/include/pybind11/detail/common.h#L493-L494

I am not sure if pybind needs to consider this a bug or even an issue. But the previous versions of pybind (2.12 and below) worked without a hitch when setting the `-Wpedantic` option, so I figured I'd let you know about this here.

### Reproducible example code

```text
#include <pybind11/pybind11.h>

int test()
{
    return 42;
}

PYBIND11_MODULE(module, m)
{
    m.def("test", &test);
}
```

### Is this a regression? Put the last known working version here if it is.

2.12.0

### PR DESCRIPTION
**Pull request by henryiii**
## Description

I think we can support `-Wpedandic` in C++20 mode. Before that the variadic macros cause warnings that are pretty much impossible to fix from a macro AFAICT. Fix #5310.

## Suggested changelog entry:

```rst
* Support `-Wpedantic` in C++20 mode
```

### CODE REVIEW

File: tests/local_bindings.h [hunk 1]
```diff
@@ -56,21 +56,21 @@ class LocalSimpleException : public std::exception {
...
-PYBIND11_MAKE_OPAQUE(LocalMap);
-PYBIND11_MAKE_OPAQUE(NonLocalVec);
-// PYBIND11_MAKE_OPAQUE(NonLocalVec2); // same type as LocalVec2
-PYBIND11_MAKE_OPAQUE(NonLocalMap);
-PYBIND11_MAKE_OPAQUE(NonLocalMap2);
+PYBIND11_MAKE_OPAQUE(LocalVec)
+PYBIND11_MAKE_OPAQUE(LocalVec2)
+PYBIND11_MAKE_OPAQUE(LocalMap)
+PYBIND11_MAKE_OPAQUE(NonLocalVec)
+// PYBIND11_MAKE_OPAQUE(NonLocalVec2)
```
Comment (henryiii) [inline tests/local_bindings.h:63]:
```suggestion
// PYBIND11_MAKE_OPAQUE(NonLocalVec2) // same type as LocalVec2
```

File: include/pybind11/detail/common.h [hunk 2]
```diff
@@ -479,6 +479,8 @@ PYBIND11_WARNING_POP
         }
 
 \endrst */
+PYBIND11_WARNING_PUSH
+PYBIND11_WARNING_DISABLE_CLANG("-Wgnu-zero-variadic-macro-arguments")
```
Comment (rwgk) [inline include/pybind11/detail/common.h:483]: Just double-checking: is this needed, even with `-std=c++20`?

**Review by rwgk — APPROVED** [review]
Thanks for the cleanup!
I think it's a great compromise, to turn on the option only for C++20. That'll will catch pretty much everything already.

File: include/pybind11/detail/common.h (hunk 2 above)
Comment (henryiii) [reply include/pybind11/detail/common.h:483]: Yeah, for some reason Clang produces a warning here. GCC is fine, but Clang complains. It produces a warnings saying this is a C++20 feature on newer compilers if set less than C++20, so no idea why the old warning triggers in C++20 mode. I think it's a bug.

**Pull request merged**